Exclude notes from excluded directories

siaw23 opened the pull request:
<!--
Thanks for contributing to Rails!

Please do not make *Draft* pull requests, as they still send
notifications to everyone watching the Rails repo.

Create a pull request when it is ready for review and feedback
from the Rails team :).

If your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

About this template

The following template aims to help contributors write a good description for their pull requests.
We'd like you to provide a description of the changes in your pull request (i.e. bugs fixed or features added), the motivation behind the changes, and complete the checklist below before opening a pull request.

Feel free to discard it if you need to (e.g. when you just fix a typo). -->

### Motivation / Background

<!--
Describe why this Pull Request needs to be merged. What bug have you fixed? What feature have you added? Why is it important?
If you are fixing a specific issue, include "Fixes #ISSUE" (replace with the issue number, remove the quotes) and the issue will be linked to this PR.
-->

The /builds folder is where compiled frontend assets are written to at build time, When we run 'rails notes' notes from this folder clutter the terminal, but these are notes from 3rd parties that we didn't write



### Additional information

<!-- Provide additional information such as benchmarks, references to other repositories, or alternative solutions. -->

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

siaw23 commented on railties/lib/rails/source_annotation_extractor.rb:
```diff
@@ -170,6 +170,7 @@ def find_in(dir)
 
       Dir.glob("#{dir}/*") do |item|
         next if File.basename(item).start_with?(".")
+        next if item.match?(%r{(^|/)app/assets/builds(/|$)})
```
Right, moved this up to avoid unnecessary work.

p8 commented on railties/lib/rails/source_annotation_extractor.rb:
```diff
@@ -168,6 +168,8 @@ def find(dirs)
     def find_in(dir)
       results = {}
 
+      return results if dir.match?(%r{(^|/)app/assets/builds(/|$)})
```
I'm wondering if this should be made more generic, there might be other directories to exclude as well.

```suggestion
      return results if dir.in?(self.class.excluded_directories)
```

This would work similar to how directories are registered:
```ruby
    class Annotation < Struct.new(:line, :tag, :text)
      def self.excluded_directories
        @@excluded_directories ||= []
      end

      def self.register_excluded_directories(*dirs)
        excluded_directories.push(*dirs)
      end
    ...
```

p8 replied:
Also, this `app/assets/builds` is configured in js-bundling-rails:
https://github.com/rails/jsbundling-rails/blob/main/lib/install/install.rb#L2-L3
Rails itself doesn't know about this directory, especially with import maps being the default.
So js-bundling-rails should probably call:
```ruby
Annotation.register_excluded_directories("app/assets/builds")
```

siaw23 replied:
@p8 this looks more elegant. I've implemented the changes and will need to check how the wiring is set up from js-bundling-rails and create a PR there too. thanks Petrik!

p8 commented on railties/test/commands/notes_test.rb:
```diff
@@ -167,6 +167,13 @@ class Rails::Command::NotesTest < ActiveSupport::TestCase
     assert_empty run_notes_command
   end
 
+  test "does not display results from generated assets under app/assets/builds" do
```
```suggestion
  test "does not display results from excluded directories" do
```

siaw23 committed:
Exclude notes from excluded directories

The /builds folder is where compiled frontend assets are written to at build time, When we run 'rails notes', notes from this folder clutter the terminal, but these are notes from 3rd parties that we didn't write
